### PR TITLE
[WIP] Clean code, add code editor

### PR DESCRIPTION
**Pull request by rubenwardy**
Soon:

* Add more pages (inventory, etc)
* Add better error messages
* Add Lua support
* Add some sort of visual support

### CODE REVIEW

**Comment by octacian** [comment]
Why bother storing the code from the editor in a physical file when it could just be stored in meta? That's my only concern/question, will test.

**Comment by rubenwardy** [comment]
Less copying when moving the node, so less node meta data sending.
It's also safer

And I'm planning on adding sharing etc

**Comment by octacian** [comment]
Only big thing I noticed when I fired it up was that you reimplemented the remote and require it for using turtle. ATM, the remote has been entirely removed until a later time. I do not think set name should be shown on place because it could get the player in a bad situation. Essentially, it's just kinda annoying.

The reason I would prefer using metadata for editor, is because it would only need to store one file, is less complex, and is more secure. Especially in the perspective of a server owner.

Turtle IDs was something I was going to do for the remote control, but it needed some work and was kinda something for v0.2.

I think what's most important right now, is a inventory and visual editor. I've already almost fully completed the editor in another branch, but it's not quite ready and is quite old.

**Comment by rubenwardy** [comment]
>  reimplemented the remote

You removed the remote after I moved it to another file
It isn't needed, I just didn't want to remove it

>  require it for using turtle

The remote is separate from everything else, if you disable remote.lua it's not added.

>  I do not think set name should be shown on place because it could get the player in a bad situation

after_build_node is always called when a player places a node, I think it's perfectly fine. It works for my usage. I don't mind disabling it though

> Turtle IDs was something I was going to do for the remote control, but it needed some work and was kinda something for v0.2

It is a rather important feature to have early before the code gets more complicated. Associating one robot per player in such a way is a bad idea

>  a inventory 

This isn't in my immediate priorities but seems useful

> and visual editor. I've already almost fully completed the editor in another branch, but it's not quite ready and is quite old.

I'd be interested in seeing this. My original plan was to experiment using a scratch alternative to generate code and then use a run button to open Minetest

**Comment by octacian** [comment]
Interesting. I will work on reimplementing the remote and turtle IDs in a separate branch, as I'd like to try it myself and get an idea. At least it's a lot less work considering I can steal some stuff from your version :stuck_out_tongue: 

I've taken a look at your latest commit, and I think that the turtle should be programmed via a tab in it's own main formspec, not via another tool. The editor your using is good, but the turtle only needs one single file. Also take a look at my [editor](https://github.com/BiLachner/TurtleMiner/tree/editor) branch.

**Comment by rubenwardy** [comment]
> I've taken a look at your latest commit, and I think that the turtle should be programmed via a tab in it's own main formspec, not via another tool

done, I was planning on doing originally (that's why I added turtleminer.on_rightclick)

> The editor your using is good, but the turtle only needs one single file

More than one file is useful though as it allows the user to have lots of different projects. I'm also planning on adding sharing etc

> Also take a look at my editor branch.

Nice attempt, still WIP though. Doesn't appear to be a visual editor. I may borrow some of the lua sandbox stuff from there

> more secure

nope. It's safely lua serialised.

> one file

I don't think this is a problem. By having multiple files less needs to be in the memory at a time. It wouldn't be difficult to make it one file though

> less complex

already implemented by a third party mod - which was designed for this purpose

--------

It seems we have different aims though - I'm aiming for a mod to teach kids how to program, you're aiming for a more realistic bot to go with digicompute

**Comment by octacian** [comment]
> already implemented by a third party mod - which was designed for this purpose

If by any means possible, I'd like to limit the amount of third-party mods required. In fact, I'd prefer none aside from those that will be already required in digicompute ([datalib](https://github.com/octacian/datalib)). I'm not trying to say that it's not good enough or something, just trying to say that more than one dependency is plain annoying...

**Comment by rubenwardy** [comment]
You can easily just include it as a subfolder. It's a library packaged by default as a mod

Just a note about my intentions: I want to make the education side of Minetest more well presented and easier to access. I compared all the available turtle mods, and none of them are good enough for my purposes - this mod is the closest to complete out of them. Thank you for all the work you've put in

I'm a great believer in reducing repeated work - would it be useful if I submitted a PR with all my clean ups and turtle ids, but without any of the controversial editor / program / script vm? That way we can share a base even if we work on different editors / "main formspecs".

The PR would include:

* Turtle IDs and saving of these
* Separate feedback (sounds) from the actual commands
* Ignore file
* Ignore node bug fixes
* Code simplification
* turtleminer.on_rightclick to try make it possible for other lua files to show things if set_name isn't required (I personally think that api.lua shouldn't depend on any other file, and the other files should just use the api. The main formspec should be in some other file IMO, to increase cohesion in api.lua - the file should just be about running turtle commands, turtle nodes, and naming the turtles)
* Misc code cleans

**Comment by octacian** [comment]
Well, then, AFAIK, you, me, and @BiLachner have essentially the same intentions.

I do now see the use of multiple files, and how what you are doing for the editor could work very well, however, it would more useful to do as you suggested and submit a PR without the editor for now. Once you've entirely completed the editor, submit a PR for that and we'll take a look. 

Considering that I am in agreement with all the work that you've done so far aside from the editor, it would work especially well. Your clean up work definitely makes it appear better over all, so it is much appreciated.

TL;DR yes, submit a PR without controversial formspec.

**Comment by octacian** [comment]
Yes, I agree. The API should contain only turtle things - node registering, naming, moving, etc... In fact, I was going to separate the formspec today, but figure I should let you prepare other PR first to avoid conflicts. Overall the changes you mentioned look good, and I would appreciate it if you would submit another PR with those. I still will have to review it in-depth and will possibly make some changes, but from what I've seen already most if not all seems fine to me.

**Edit:** I'm beginning to see a few things that would need to be changed before merged, but I'll leave that until you open the other PR so as to keep discussion on topic.
